src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,692 @@
+            AstBuilder astBuilder, 
+            IEnumerable<NodeModel> graph,
+            IEnumerable<NodeModel> nodes)
         {
-            var astNodes = astBuilder.CompileToAstNodes(nodeList, false, verboseLogging);
-            var codeGen = new ProtoCore.CodeGenDS(astNodes);
-            return codeGen.GenerateCode();
+            // The basic worflow is:
+            //   1. Compile each node to get its cde in AST format
+            // 
+            //   2. Variable numbering to avoid confliction. For example, two 
+            //      code block nodes both have assignment "y = x", we need to 
+            //      rename "y" to "y1" and "y2" respectively.
+            //
+            //   3. Variable remapping. For example, code block node 
+            //      "x = 1" connects to "a = b", where the second code block 
+            //      node will have initialization "b = x_guid" where "x_guid"
+            //      is because of variable mappining in the first code block 
+            //      node. We should restore it to its original name.
+            //   
+            //      Note in step 2 we may rename "x" to "xn". 
+            //
+            //   4. Generate short name for long name variables. Typically they
+            //      are from output ports from other nodes.
+            //
+            //   5. Do constant progation to optimize the generated code.
+            #region Step 1 AST compilation
+
+            var sortedGraph = AstBuilder.TopologicalSortForGraph(graph);
+            var sortedNodes = sortedGraph.Where(nodes.Contains);
+
+            var allAstNodes = astBuilder.CompileToAstNodes(sortedNodes, AstBuilder.CompilationContext.NodeToCode, false);
+
+            #endregion
+
+            #region Step 2 Varialbe numbering
+            // In this step, we'll renumber all variables that going to be in 
+            // the same code block node. That is, 
+            // 
+            //     "x = 1; y = x;"   and
+            //     "x = 2; y = x;" 
+            //
+            // Will be renumbered to 
+            //
+            //    "x1 = 1; y1 = x1;" and
+            //    "x2 = 2; y2 = x2;"
+
+            // Map from mapped variable to its original name. These variables 
+            // are from code block nodes that in the selection.
+            Dictionary<string, string> renamingMap = null; ;
+
+            // Input variable to renamed input variable map
+            Dictionary<string, string> inputMap = null;
+
+            // Output variable to renamed output variable map
+            Dictionary<string, string> outputMap = null;
+
+            // Collect all inputs/outputs/candidate renaming variables
+            GetInputOutputMap(nodes, out inputMap, out outputMap, out renamingMap);
+
+            // Vairable numbering map. The Tuple value indicates its number
+            // sequence and if for the new UI node.
+            var numberingMap = new Dictionary<string, NumberingState>();
```

Comment:
I think the comment here *The Tuple value indicates...* might be slightly outdated?